database/src/service/read.rs

```diff
@@ -228,57 +342,177 @@ fn map_request(
+    let table_block_heights = env_inner.open_db_ro::<BlockHeights>(&tx_ro)?;
+    let table_block_infos = env_inner.open_db_ro::<BlockInfos>(&tx_ro)?;
+
+    let top_height = top_block_height(&table_block_heights)?;
+
+    Ok(Response::GeneratedCoins(cumulative_generated_coins(
+        &top_height,
+        &table_block_infos,
+    )?))
 }
 
 /// [`ReadRequest::Outputs`].
 #[inline]
-#[allow(clippy::needless_pass_by_value)] // TODO: remove me
-fn outputs(env: &Arc<ConcreteEnv>, map: HashMap<Amount, HashSet<AmountIndex>>) -> ResponseResult {
-    todo!()
+fn outputs(env: &ConcreteEnv, map: HashMap<Amount, HashSet<AmountIndex>>) -> ResponseResult {
+    let env_inner = env.env_inner();
+    let tx_ro = set_tx_ro!(env, env_inner);
+    let tables = set_tables!(env, env_inner, tx_ro);
+
+    // -> Result<(AmountIndex, OutputOnChain), RuntimeError>
+    let inner_map = |amount, amount_index| {
+        let (tx_ro, tables) = get_tx_ro_and_tables!(env_inner, tx_ro, tables);
+
+        let pre_rct_output_id = PreRctOutputId {
+            amount,
+            amount_index,
+        };
+
+        let output = get_output(&pre_rct_output_id, tables.outputs())?;
+        let output_on_chain = output_to_output_on_chain(&output, amount, tables.tx_unlock_time())?;
+
+        Ok((amount_index, output_on_chain))
+    };
+
+    let map = map
+        .into_par_iter()
+        .map(|(amount, amount_index_set)| {
+            Ok((
+                amount,
+                amount_index_set
+                    .into_par_iter()
+                    .map(|amount_index| inner_map(amount, amount_index))
+                    .collect::<Result<HashMap<AmountIndex, OutputOnChain>, RuntimeError>>()?,
+            ))
+        })
+        .collect::<Result<HashMap<Amount, HashMap<AmountIndex, OutputOnChain>>, RuntimeError>>()?;
+
+    Ok(Response::Outputs(map))
 }
 
 /// [`ReadRequest::NumberOutputsWithAmount`].
-/// TODO
 #[inline]
-#[allow(clippy::needless_pass_by_value)] // TODO: remove me
-fn number_outputs_with_amount(env: &Arc<ConcreteEnv>, vec: Vec<Amount>) -> ResponseResult {
-    todo!()
+fn number_outputs_with_amount(env: &ConcreteEnv, amounts: Vec<Amount>) -> ResponseResult {
+    let env_inner = env.env_inner();
+    let tx_ro = set_tx_ro!(env, env_inner);
+    let tables = set_tables!(env, env_inner, tx_ro);
+
+    let map = amounts
+        .into_par_iter()
+        .map(|amount| {
+            let (tx_ro, tables) = get_tx_ro_and_tables!(env_inner, tx_ro, tables);
+
+            match tables.num_outputs().get(&amount) {
+                // INVARIANT: #[cfg] @ lib.rs asserts `usize == u64`
+                #[allow(clippy::cast_possible_truncation)]
+                Ok(count) => Ok((amount, count as usize)),
+                // If we get a request for an `amount` that doesn't exist,
+                // we return `0` instead of an error.
+                Err(RuntimeError::KeyNotFound) => Ok((amount, 0)),
+                Err(e) => Err(e),
+            }
```

Comment:
I think we need to handle RCT outputs here as well

Reply:
If RCT, what number are we supposed to return? The total amount of RCT outputs? `tables.rct_outputs().len()`?

Reply:
yes total amount of RCT outputs